lib/src/back_end/solver.dart

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+import 'package:collection/collection.dart';
+
+import '../dart_formatter.dart';
+import '../debug.dart' as debug;
+import '../piece/piece.dart';
+import 'solution.dart';
+
+/// Selects states for each piece in a tree of pieces to find the best set of
+/// line splits that minimizes overflow characters and line splitting costs.
+///
+/// This problem is combinatorial for the number of pieces and each of their
+/// possible states, so it isn't feasible to brute force. There are a few
+/// techniques we use to avoid that:
+///
+/// -   All pieces default to being in state zero. Every piece is implemented
+///     such that that state has no line splits and zero cost. Thus, it tries
+///     solutions with a minimum number of line splits first.
+///
+/// -   Solutions are explored in priority order. We explore solutions with the
+///     fewest overflow characters and the lowest cost (in that order) first.
+///     This was, as soon as we find a solution with no overflow characters, we
+///     know it will be the best solution and can stop.
+// TODO(perf): At some point, we probably also want to do memoization of
+// previously formatted Piece subtrees.
+class Solver {
+  final DartFormatter _formatter;
```

Comment:
It looks like we only need the `pageWidth` field from `_formatter` - can we store that directly instead of the DartFormatter?

Reply:
Oh good catch. Done.